File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # 1,2,3,1,2,3
        #     i---j

        freq = {}
        if k==0: return False
        i, j = 0, 0
        while j < len(nums):
            if nums[j] in freq:
                return True
            if j - i >= k:
                freq.pop(nums[i])
                i += 1
            freq[nums[j]] = 1
            j += 1

        return False

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashSet<Integer> seen = new HashSet<>();
        
        // expand the window
        int l = 0, r = 0;
        while (r < nums.length && r - l <= k) {
            if (seen.contains(nums[r])) return true;
            seen.add(nums[r]);
            r++;
        }

        // slide the window
        while (r < nums.length) {
            seen.remove(nums[l++]);
            if (seen.contains(nums[r])) return true;
            seen.add(nums[r]);
            r++;
        }

        return false;   
    }
}